types/rlp_unmarshal.go

```diff
@@ -376,44 +384,62 @@ func (t *Transaction) unmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
 		return err
 	}
 
-	if len(elems) < 9 {
-		return fmt.Errorf("incorrect number of elements to decode transaction, expected 9 but found %d", len(elems))
+	getElem := func() *fastrlp.Value {
+		val := elems[0]
+		elems = elems[1:]
+
+		return val
+	}
+
+	var num int
+
+	switch t.Type {
+	case LegacyTx:
+		num = 9
+	case DynamicFeeTx, StateTx:
+		num = 10
+	default:
+		return fmt.Errorf("transaction type %d not found", t.Type)
+	}
+
+	if numElems := len(elems); numElems != num {
+		return fmt.Errorf("incorrect number of transaction elements, expected %d but found %d", num, numElems)
 	}
 
 	p.Hash(t.Hash[:0], v)
 
 	// nonce
-	if t.Nonce, err = elems[0].GetUint64(); err != nil {
+	if t.Nonce, err = getElem().GetUint64(); err != nil {
```

Comment:
Correct me if I'm wrong, but this getElem will not return the correct values if this t.Type == DynamicFeeTx condition is met, because it goes from the first element in array and decreases the array size by each call, and we were extracting the 10th and 11th index from the array before.